File: scripts/preprocessing/gee/python/landsat_time_series_to_poly.py
# ---
# title:   Summarize Landsat Time Series to Polygons
# author:  Brendan Casey
# created: 2026-07-10
# inputs:
#   - Landsat 5/7/8/9 Surface Reflectance collections
#     (LANDSAT/*/C02/T1_L2)
#   - Summary polygons (test FeatureCollection)
#   - FAO GAUL province boundaries (Alberta)
# outputs:
#   - Per-polygon-per-date spectral-index summary CSV
#     (ls_poly_summary) exported to Google Drive
# notes:
#   Python port of landsat_time_series_to_poly.js for the
#   Earth Engine Python API. Builds an annual date list,
#   computes user-selected spectral indices via the shared
#   ls_fn helper, drops the QA_PIXEL band, casts to
#   Float32, and reduces each image over polygons with
#   image_collection_to_features, exporting the result as a
#   CSV.
#
#   The original Map.addLayer calls and debug print()
#   blocks are omitted.
#
#   Deviation: the shared ls_fn helper expects a client-
#   side list of date strings, so the ee.List produced by
#   create_date_list is materialized with getInfo() before
#   being passed in.
#
#   Setup (once):
#     pip install earthengine-api
#     earthengine authenticate
#   Then set EE_PROJECT in _gee_config.py and run.
# ---

import os
import sys

import ee

# Make utils importable regardless of the working
# directory VS Code runs the script from
sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))

from utils.compute_report import ComputeReport
from utils.gee_helpers import create_date_list
from utils.gee_utils import initialize_ee
from utils.image_collection_to_features import (
    image_collection_to_features,
)
from utils.landsat_time_series import ls_fn

# 1. Setup ----

# 1.1 User parameters ----
LS_START_DATE = "2009-06-01"  # first time-series date
LS_END_DATE = "2024-06-01"  # last time-series date
LS_DATE_INTERVAL = 1  # step between series start dates
LS_DATE_INTERVAL_TYPE = "years"  # units for the step
LS_WINDOW = 121  # compositing window length
LS_WINDOW_TYPE = "days"  # units for the window
LS_STATISTIC = "mean"  # 'mean', 'median', 'max', etc.
LS_INDICES = [
    "BSI", "DRS", "DSWI", "EVI", "GNDVI",
    "LAI", "NBR", "NDMI", "NDSI", "NDVI",
    "NDWI", "SAVI", "SI",
]
SUMMARY_SCALE = 30  # reduction scale (m)
SUMMARY_CRS = "EPSG:4326"  # reduction CRS
SUMMARY_TILE_SCALE = 4  # tileScale for parallel reduction
SUMMARY_FILE_NAME = "ls_poly_summary"  # output CSV prefix
PRINT_STATS = True  # summary check (slow for large AOIs)
USE_TEST_AOI = True  # True: small test AOI; False: Alberta
COMPUTE_REPORT = True  # write EECU usage report (txt)

# 1.2 Initialize Earth Engine ----
# Project ID is read from _gee_config.py
initialize_ee()

# 1.3 Set up compute usage report ----
# Profiles EECU usage per section. Best used with
# USE_TEST_AOI = True to find choke points cheaply.
report = ComputeReport(
    "landsat_time_series_to_poly",
    out_dir=os.path.join(
        os.path.dirname(os.path.abspath(__file__)),
        "gee_compute_reports",
    ),
    enabled=COMPUTE_REPORT,
)

# 2. Define study area ----
# Uses a small test polygon when USE_TEST_AOI is True;
# otherwise filters the FAO GAUL provinces for Alberta.

if USE_TEST_AOI:
    # Small aoi for testing purposes
    aoi = ee.Geometry.Polygon([
        [-113.5, 55.5],  # Top-left corner
        [-113.5, 55.0],  # Bottom-left corner
        [-112.8, 55.0],  # Bottom-right corner
        [-112.8, 55.5],  # Top-right corner
    ])
else:
    aoi = (
        ee.FeatureCollection(
            "FAO/GAUL_SIMPLIFIED_500m/2015/level1"
        )
        .filter(ee.Filter.eq("ADM0_NAME", "Canada"))
        .filter(ee.Filter.eq("ADM1_NAME", "Alberta"))
        .geometry()
    )

# 2.1 Define summary polygons ----
# Five ~2 ha test polygons within the AOI, each tagged with
# an integer id.
poly1 = ee.Geometry.Polygon([
    [-113.48, 55.48],
    [-113.48, 55.47873],
    [-113.47779, 55.47873],
    [-113.47779, 55.48],
])
poly2 = ee.Geometry.Polygon([
    [-113.46, 55.47],
    [-113.46, 55.46873],
    [-113.45779, 55.46873],
    [-113.45779, 55.47],
])
poly3 = ee.Geometry.Polygon([
    [-113.44, 55.46],
    [-113.44, 55.45873],
    [-113.43779, 55.45873],
    [-113.43779, 55.46],
])
poly4 = ee.Geometry.Polygon([
    [-113.42, 55.45],
    [-113.42, 55.44873],
    [-113.41779, 55.44873],
    [-113.41779, 55.45],
])
poly5 = ee.Geometry.Polygon([
    [-113.40, 55.44],
    [-113.40, 55.43873],
    [-113.39779, 55.43873],
    [-113.39779, 55.44],
])

polys_fc = ee.FeatureCollection([
    ee.Feature(poly1, {"id": 1}),
    ee.Feature(poly2, {"id": 2}),
    ee.Feature(poly3, {"id": 3}),
    ee.Feature(poly4, {"id": 4}),
    ee.Feature(poly5, {"id": 5}),
])

# 3. Build the time-series date list ----
# create_date_list returns an ee.List; ls_fn iterates a
# client-side list, so the dates are materialized as
# YYYY-MM-dd strings.
date_list = create_date_list(
    ee.Date(LS_START_DATE),
    ee.Date(LS_END_DATE),
    LS_DATE_INTERVAL,
    LS_DATE_INTERVAL_TYPE,
)
start_dates = (
    date_list.map(
        lambda d: ee.Date(d).format("YYYY-MM-dd")
    ).getInfo()
)

# 4. Landsat time-series processing ----
# Computes the selected spectral indices for each interval,
# drops the QA_PIXEL band, and casts every band to Float32.
ls = ls_fn(
    start_dates,
    LS_WINDOW,
    LS_WINDOW_TYPE,
    aoi,
    LS_INDICES,
    LS_STATISTIC,
)


def drop_qa_and_cast(image):
    """Drop the QA_PIXEL band and cast bands to Float32."""
    keep = image.bandNames().filter(
        ee.Filter.neq("item", "QA_PIXEL")
    )
    return image.select(keep).toFloat()


ls = ls.map(drop_qa_and_cast)

# 4.1 Check calculated bands (optional) ----
# Earth Engine is lazy, so the profiler needs an evaluated
# computation to measure per-algorithm EECU usage.
if PRINT_STATS or COMPUTE_REPORT:
    with report.section("Landsat band names"):
        band_names = ls.first().bandNames().getInfo()
    print("Landsat band names:", band_names)

# 5. Summarize Landsat time series to polygons ----
# Applies the mean reducer to every band of each image over
# each polygon, producing a per-polygon-per-date summary
# table that is exported to Google Drive as a CSV.
ls_poly_summary = image_collection_to_features(
    ee.Reducer.mean(),
    polys_fc,
    aoi,
    ls,
    SUMMARY_CRS,
    SUMMARY_SCALE,
    SUMMARY_TILE_SCALE,
    SUMMARY_FILE_NAME,
)

# 6. Compute usage report ----
# Writes the profiled sections to gee_compute_reports/. The
# summary export runs as a batch table task; monitor
# progress at https://code.earthengine.google.com/tasks
report.write()

# End of script ----

File: scripts/preprocessing/gee/javascript/landsat_time_series_to_poly.js
/*
 * ---
 * title: "Summarize Landsat Time Series to Polygons"
 * author: "Brendan Casey"
 * created: "2025-08-22"
 * description: Generates a time series of Landsat satellite imagery,
 * calculates user-defined spectral indices, and summarize the to polygons. 
 * ---
 */

/* 1. Setup
 * Prepare the environment, including the AOI, helper functions,
 * and date list for time series processing.
 */ 

/* Load helper functions */
var utils = require(
  "users/bgcasey/science_centre:functions/utils"
  );
var landsatTimeSeries = require(
  "users/bgcasey/science_centre:functions/landsat_time_series"
  );
var landsatIndicesAndMasks = require(
  "users/bgcasey/science_centre:functions/landsat_indices_and_masks"
  );
var summarize = require(
  "users/bgcasey/science_centre_dev:functions/image_collection_to_features"
  );  

/* Define area of interest (AOI) */
// var aoi = ee.FeatureCollection('FAO/GAUL_SIMPLIFIED_500m/2015/level1')
//   .filter(ee.Filter.eq('ADM0_NAME', 'Canada'))
//   .filter(ee.Filter.eq('ADM1_NAME', 'Alberta'))
//   .geometry()

/* Small aoi for testing purposes */
var aoi = ee.Geometry.Polygon([
  [-113.5, 55.5],  // Top-left corner
  [-113.5, 55.0],  // Bottom-left corner
  [-112.8, 55.0],  // Bottom-right corner
  [-112.8, 55.5]   // Top-right corner
]);

/* Create polygons for testing */
// Five smaller polygons within AOI (approx. 2 ha each)
var poly1 = ee.Geometry.Polygon([
  [-113.48, 55.48],
  [-113.48, 55.47873],
  [-113.47779, 55.47873],
  [-113.47779, 55.48]
]);

var poly2 = ee.Geometry.Polygon([
  [-113.46, 55.47],
  [-113.46, 55.46873],
  [-113.45779, 55.46873],
  [-113.45779, 55.47]
]);

var poly3 = ee.Geometry.Polygon([
  [-113.44, 55.46],
  [-113.44, 55.45873],
  [-113.43779, 55.45873],
  [-113.43779, 55.46]
]);

var poly4 = ee.Geometry.Polygon([
  [-113.42, 55.45],
  [-113.42, 55.44873],
  [-113.41779, 55.44873],
  [-113.41779, 55.45]
]);

var poly5 = ee.Geometry.Polygon([
  [-113.40, 55.44],
  [-113.40, 55.43873],
  [-113.39779, 55.43873],
  [-113.39779, 55.44]
]);

// Create features for each polygon
var feature1 = ee.Feature(poly1, {id: 1});
var feature2 = ee.Feature(poly2, {id: 2});
var feature3 = ee.Feature(poly3, {id: 3});
var feature4 = ee.Feature(poly4, {id: 4});
var feature5 = ee.Feature(poly5, {id: 5});

// Combine features into a single FeatureCollection
var polys_fc = ee.FeatureCollection([
  feature1,
  feature2,
  feature3,
  feature4,
  feature5
]);


/* Create a date list 
 * The date list specifies the starting points for time 
 * intervals used to extract a time series. The createDateList 
 * function generates a list of dates at a specified interval 
 * (e.g., 1 year), beginning on the provided start date 
 * ('2000-06-01') and ending on the end date ('2024-06-01').
 *
 * For each date in the list, the ls_fn function will create a 
 * new end date by advancing the start date by a user-defined 
 * number of time units (e.g., 4 months, 6 weeks). Indices will 
 * be calculated for each of these time intervals.
 *
 * Due to memory limits when generating a time series of
 * Alberta wide images, time series are generated in five year
 * batches. Comment out the unused time periods.  
 */

var dateList = utils.createDateList(
  ee.Date('2009-06-01'), ee.Date('2024-06-01'), 1, 'years'
);
 
print("Start Dates", dateList);

/* Define reducer statistic */
var statistic = 'mean'; // Choose from 'mean', 'median', 'max', etc.


/* 2. Landsat Time Series Processing
 * Calculate user-defined spectral indices for Landsat imagery.
 * 
 * Available indices:
 * - BSI: Bare Soil Index
 * - DRS: Distance Red & SWIR
 * - DSWI: Disease Stress Water Index
 * - EVI: Enhanced Vegetation Index
 * - GNDVI: Green Normalized Difference Vegetation Index
 * - LAI: Leaf Area Index
 * - NBR: Normalized Burn Ratio
 * - NDMI: Normalized Difference Moisture Index
 * - NDSI: Normalized Difference Snow Index
 * - NDVI: Normalized Difference Vegetation Index
 * - NDWI: Normalized Difference Water Index
 * - SAVI: Soil Adjusted Vegetation Index
 * - SI: Shadow Index
 */
var ls = landsatTimeSeries.ls_fn(
  dateList, 121, 'days', aoi,
  [
    'BSI', 'DRS', 'DSWI', 'EVI', 'GNDVI', 
    'LAI', 'NBR', 'NDMI', 'NDSI', 'NDVI',
    'NDWI', 'SAVI', 'SI'
  ],
  statistic
)
  .map(function(image) {
    // Exclude QA_PIXEL band and rename remaining bands
    var filteredBandNames = image.bandNames().filter(
      ee.Filter.neq('item', 'QA_PIXEL')
    );
    return image
      .select(filteredBandNames)
      .toFloat(); // Convert all bands to Float32
  });

//print("Landsat Time Series:", ls);

/* 3. Summarize Landsat Time Series to Polygons
 * This section extracts summary statistics from the Landsat 
 * time series (ls) over each polygon in a feature collection.
 * The function imageCollectionToFeatures applies the specified 
 * reducer to all bands in each landsat image for every polygon, 
 * producing a per-polygon-per-date summary table.
 * 
 * The results are exported to Google Drive as a CSV file 
 * (one row per polygon-date-band combination).
 */

var ls_poly_summary = summarize.imageCollectionToFeatures(
  ee.Reducer.mean(),    // Reducer: mean value in each polygon
  polys_fc,             // FeatureCollection: 2-ha polygons
  aoi,                  // AOI (for filtering polygons)
  ls,                   // Landsat time series collection
  'EPSG:4326',          // CRS
  30,                   // Scale in meters
  4,                    // tileScale
  'ls_poly_summary'     // Output file name prefix
);



